jerry-ext/include/jerryscript-ext/args.h

```diff
@@ -0,0 +1,99 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef JERRYX_ARGS_H
+#define JERRYX_ARGS_H
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include "jerryscript.h"
+
+typedef struct jerryx_arg_t jerryx_arg_t;
+typedef struct jerryx_arg_iterator_t jerryx_arg_iterator_t;
+
+/**
+ * Signature of the transform funtion.
+ */
+typedef jerry_value_t (*jerryx_arg_transform_func_t) (jerryx_arg_iterator_t *iterator_p);
+
+/**
+ * The structure of the jerry argument in arg transform.
+ * Each native argument should have one corresponding jerryx_arg_t item.
+ */
+struct jerryx_arg_t
+{
+  jerryx_arg_transform_func_t func; /**< the transform function for the native argument */
+  void *dest; /**< point to the native argument */
+  uintptr_t extra_info; /**< extra infomation for arg transform */
+};
+
+/**
+ * The argument iteration used for transform funtions.
+ */
+struct jerryx_arg_iterator_t
+{
+  const jerryx_arg_t *c_arg_p; /**< points to the current jerryx_arg_t */
+  const jerry_value_t *js_arg_p; /**< points to the current js argument */
+  const jerry_length_t  js_arg_cnt; /**< the total count of the js arguments */
+  jerry_length_t  js_arg_index; /**< the current index of the js argument */
+};
+
+jerry_value_t jerryx_validate_and_assign_all (const jerry_value_t this_val,
+                                              const jerry_value_t *js_arg_p,
+                                              const jerry_length_t js_arg_cnt,
+                                              const jerryx_arg_t *c_arg_p,
+                                              const jerry_length_t c_arg_cnt);
+
+jerry_value_t jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p,
+                                               const jerry_length_t js_arg_cnt,
+                                               const jerryx_arg_t *c_arg_p,
+                                               const jerry_length_t c_arg_cnt);
+
+/* Helper functions for jerryx_arg_iterator_t. */
+jerry_value_t jerryx_arg_iterator_pop (jerryx_arg_iterator_t *iterator_p);
+void *jerryx_arg_iterator_get_dest (jerryx_arg_iterator_t *iterator_p);
+
+/**
+ * Indicates whether a argument can be converted into the target type.
+ */
+typedef enum
+{
+  JERRY_ARG_CONVERSION_YES, /**< the transform inside will invoke toNumber, toBoolean or toString */
+  JERRY_ARG_CONVERSION_NO /**< the conversion is not allowed. */
```

Comment:
Should these also be prefixed with `JERRYX_...` ?

I'd rather use "coerce" instead of "conversion".
I think this word is usually used for automatic type conversion in JS?
`JERRYX_ARG_COERCE` vs `JERRYX_ARG_NO_COERCE` ?

Also note with the "no" option that the validation will not pass if the value isn't the expected type.